### PR TITLE
analyzer summary bug fixes

### PR DESCRIPTION
**Pull request by bsdangra**

### CODE REVIEW

File: soiapis/utils.py [hunk 1]
```diff
@@ -688,11 +691,31 @@ def get_analyzer_data(countries, brands, fy_year):
...
+    temp_master_obj = {}
+
     for i in soi_base_data:
-        if not i.model_name in unique_model_name_list: unique_model_name_list.append(i.model_name)
-    adstock_master_obj=get_adstock_master_obj(unique_model_name_list,start_date,end_date,fy_year)
-    key_list=list(adstock_master_obj.keys())
+        if not i.model_name in unique_model_name_list: 
+            unique_model_name_list.append(i.model_name)
+            temp_master_obj[i.model_name] = {
+                "base_data" : SoiBaseDetails.objects.filter(model_name=i.model_name, date__range=[start_date, end_date]),
```
Comment (nabendukarmakar) [inline soiapis/utils.py]: Why do we need to do a DB call again? Isn't the data already available?

File: soiapis/utils.py [hunk 2]
```diff
@@ -688,11 +691,31 @@ def get_analyzer_data(countries, brands, fy_year):
...
     unique_model_name_list = []
+    temp_master_obj = {}
+
     for i in soi_base_data:
-        if not i.model_name in unique_model_name_list: unique_model_name_list.append(i.model_name)
-    adstock_master_obj=get_adstock_master_obj(unique_model_name_list,start_date,end_date,fy_year)
-    key_list=list(adstock_master_obj.keys())
+        if not i.model_name in unique_model_name_list: 
+            unique_model_name_list.append(i.model_name)
+            temp_master_obj[i.model_name] = {
```
Comment (nabendukarmakar) [inline soiapis/utils.py]: why is it called `temp_master_obj`?

File: soiapis/utils.py [hunk 3]
```diff
@@ -688,11 +691,31 @@ def get_analyzer_data(countries, brands, fy_year):
...
+    key_list = list(adstock_master_obj.keys())
+
+    for model in temp_master_obj:
+        temp_base_obj = temp_master_obj[model]["base_data"]
+        temp_mapped_keys = []
+        for i in temp_base_obj:
+            temp_mapped_keys.append(i.mapped_key)
+        temp_mapped_keys = list(set(temp_mapped_keys))
+        temp_master_obj[model]["kbd_data"] = SoiKbdDetails.objects.filter(mapped_key__in=temp_mapped_keys)
+        temp_master_obj[model]["op_data"] = get_op_data(temp_master_obj[model]["base_data"], model, temp_master_obj[model]["kbd_data"])
```
Comment (nabendukarmakar) [inline soiapis/utils.py]: This also makes a DB Call I believe.

Need to reduce the DB calls and no DB calls inside a For Loop.

File: soiapis/utils.py [hunk 4]
```diff
@@ -688,11 +691,31 @@ def get_analyzer_data(countries, brands, fy_year):
...
+    adstock_master_obj = get_adstock_master_obj(unique_model_name_list,start_date,end_date,fy_year)
+    key_list = list(adstock_master_obj.keys())
+
+    for model in temp_master_obj:
+        temp_base_obj = temp_master_obj[model]["base_data"]
+        temp_mapped_keys = []
+        for i in temp_base_obj:
+            temp_mapped_keys.append(i.mapped_key)
+        temp_mapped_keys = list(set(temp_mapped_keys))
+        temp_master_obj[model]["kbd_data"] = SoiKbdDetails.objects.filter(mapped_key__in=temp_mapped_keys)
```
Comment (nabendukarmakar) [inline soiapis/utils.py]: No DB Call in For Loop Please.

File: soiapis/utils.py (hunk 3 above)
Comment (bsdangra) [reply soiapis/utils.py]: This db call is required. The number of db call is reduced significantly compared to previous version.

File: soiapis/utils.py (hunk 2 above)
Comment (bsdangra) [reply soiapis/utils.py]: This is dict, contains model level data. Will change the name.

File: soiapis/utils.py (hunk 1 above)
Comment (bsdangra) [reply soiapis/utils.py]: Will use the available data.

File: soiapis/utils.py [hunk 5]
```diff
@@ -688,28 +691,48 @@ def get_analyzer_data(countries, brands, fy_year):
...
+    uniq_master_obj_dict = {}
+
     for i in soi_base_data:
-        if not i.model_name in unique_model_name_list: unique_model_name_list.append(i.model_name)
-    adstock_master_obj=get_adstock_master_obj(unique_model_name_list,start_date,end_date,fy_year)
-    key_list=list(adstock_master_obj.keys())
+        if not i.model_name in unique_model_name_list: 
+            unique_model_name_list.append(i.model_name)
+            uniq_master_obj_dict[i.model_name] = {
+                "base_data" : soi_base_data.filter(model_name=i.model_name, date__range=[start_date, end_date]),
```
Comment (nabendukarmakar) [inline soiapis/utils.py]: will it make a DB call?

File: soiapis/utils.py [hunk 6]
```diff
@@ -688,28 +691,48 @@ def get_analyzer_data(countries, brands, fy_year):
...
-    len(kbd_data)
+
+    for model in uniq_master_obj_dict:
+        temp_base_obj = uniq_master_obj_dict[model]["base_data"]
+        temp_mapped_keys = []
+        for i in temp_base_obj:
+            temp_mapped_keys.append(i.mapped_key)
+        temp_mapped_keys = list(set(temp_mapped_keys))
+        uniq_master_obj_dict[model]["kbd_data"] = kbd_data.filter(mapped_key__in=temp_mapped_keys)
+        uniq_master_obj_dict[model]["op_data"] = get_op_data(uniq_master_obj_dict[model]["base_data"], model, uniq_master_obj_dict[model]["kbd_data"])
```
Comment (nabendukarmakar) [inline soiapis/utils.py]: `get_op_data` makes DB calls right?

Comment (bsdangra) [reply soiapis/utils.py]: No Nabendu, no db call in get_op_data func

**Pull request merged**